// dbots/lib/Interface/Lists/BladeList.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var Service_1 = require("../Service");
var Util_1 = require("../../Utils/Util");
/**
 * Represents the BladeList service.
 * @see https://docs.bladelist.gg/en/latest/api/index.html
 */
var BladeList = /** @class */ (function (_super) {
    __extends(BladeList, _super);
    function BladeList() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(BladeList, "aliases", {
        /** The values that can be used to select the service. */
        get: function () {
            return ['bladebotlist', 'bladebotlist.xyz', 'bladelist', 'bladelist.gg'];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BladeList, "logoURL", {
        /** The logo URL. */
        get: function () {
            return 'https://bladelist.gg/static/img/logo.png';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BladeList, "serviceName", {
        /** Service's name. */
        get: function () {
            return 'BladeList';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BladeList, "websiteURL", {
        /** The website URL. */
        get: function () {
            return 'https://bladelist.gg';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BladeList, "baseURL", {
        /** The base URL of the service's API. */
        get: function () {
            return 'https://api.bladelist.gg';
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Posts statistics to this service.
     * @param options The options of the request.
     */
    BladeList.post = function (options) {
        var token = options.token, clientID = options.clientID, serverCount = options.serverCount, shard = options.shard;
        return _super._post.call(this, {
            method: 'put',
            url: "/bots/" + Util_1.Util.resolveID(clientID) + "/",
            headers: {
                Authorization: token,
                'Content-Type': 'application/json'
            },
            data: __assign({ server_count: Util_1.Util.resolveCount(serverCount) }, (shard ? { shard_count: shard.count } : {}))
        });
    };
    /**
     * Gets a bot listed on this service.
     * @param id The bot's ID.
     */
    BladeList.prototype.getBot = function (id) {
        return this._request({
            url: "/bots/" + Util_1.Util.resolveID(id),
            headers: { Authorization: this.token }
        }, { requiresToken: true });
    };
    return BladeList;
}(Service_1.Service));
exports.default = BladeList;
